pkg/controller/direct/iam/partialpolicy_controller.go

```diff
@@ -0,0 +1,403 @@
+				Name:      memberFrom.LogSinkRef.Name,
+				Namespace: memberFrom.LogSinkRef.Namespace,
+			}
+		}
+		if memberFrom.SQLInstanceRef != nil {
+			oldMemberFrom.SQLInstanceRef = &oldiamv1beta1.MemberReference{
+				Name:      memberFrom.SQLInstanceRef.Name,
+				Namespace: memberFrom.SQLInstanceRef.Namespace,
+			}
+		}
+		if memberFrom.ServiceIdentityRef != nil {
+			oldMemberFrom.ServiceIdentityRef = &oldiamv1beta1.MemberReference{
+				Name:      memberFrom.ServiceIdentityRef.Name,
+				Namespace: memberFrom.ServiceIdentityRef.Namespace,
+			}
+		}
+
+	}
+
+	return kcciamclient.ResolveMemberIdentity(t.Ctx, oldMember, oldMemberFrom, defaultNamespace, t.Iamclient.TFIAMClient)
+}
+
+func (a *IAMPartialPolicyAdapter) Create(ctx context.Context, createOp *directbase.CreateOperation) error {
+	log := getLogger(ctx)
+	log.V(2).Info("creating/applying IAMPartialPolicy", "name", k8s.GetNamespacedName(a.desired))
+
+	var livePolicyForMerge *newiamv1beta1.IAMPolicy
+
+	// If the policy truly doesn't exist, livePolicyForMerge should be an empty IAMPolicy struct
+	// for ComputePartialPolicyWithMergedBindings to work correctly.
+	if a.actualReferencedResourcePolicy == nil {
+		livePolicyForMerge = &newiamv1beta1.IAMPolicy{}
+	} else {
+		// todo acpana round trip ?
+		livePolicyForMerge = ToNewIAMPolicySkeleton(a.desired)
+	}
+
+	resolver := IAMMemberIdentityResolver{Iamclient: a.iamClient, Ctx: ctx}
+	desiredPartialPolicyWithStatus, err := ComputePartialPolicyWithMergedBindings(a.desired, livePolicyForMerge, resolver)
+	if err != nil {
+		return fmt.Errorf("computing partial policy for create for %v: %w", k8s.GetNamespacedName(a.desired), err)
+	}
+
+	finalPolicyToSet := toDesiredPolicy(desiredPartialPolicyWithStatus, livePolicyForMerge)
+
+	_, err = a.iamClient.SetPolicy(ctx, finalPolicyToSet)
+	if err != nil {
+		return fmt.Errorf("setting IAM policy for %v: %w", k8s.GetNamespacedName(a.desired), err)
+	}
+
+	log.V(2).Info("successfully applied IAMPartialPolicy for create", "name", k8s.GetNamespacedName(a.desired))
+
+	// Update KRM status
+	newKRMStatus := newiamv1beta1.IAMPartialPolicyStatus{
+		ObservedGeneration:  a.desired.GetGeneration(),
+		LastAppliedBindings: desiredPartialPolicyWithStatus.Status.LastAppliedBindings,
+		AllBindings:         desiredPartialPolicyWithStatus.Status.AllBindings,
+	}
+	readyCondition := k8s.NewCustomReadyCondition(corev1.ConditionTrue, k8s.UpToDate, k8s.UpToDateMessage)
+	if err := createOp.UpdateStatus(ctx, &newKRMStatus, &readyCondition); err != nil {
+		return fmt.Errorf("updating status for %v: %w", k8s.GetNamespacedName(a.desired), err)
+	}
+	return nil
+}
+
+func (a *IAMPartialPolicyAdapter) Update(ctx context.Context, updateOp *directbase.UpdateOperation) error {
+	log := getLogger(ctx)
+	log.V(2).Info("updating/applying IAMPartialPolicy", "name", k8s.GetNamespacedName(a.desired))
+	mapCtx := &direct.MapContext{}
+
+	if a.actualReferencedResourcePolicy == nil {
```

Comment:
IIUC, when the actual policy is nil, Create() should be triggered and this branch should never be hit.

Though in the Find() option, it returns true regardless if policy is nil or not. Looks like the users can never create policy when the referenced resource exist but the IAM policy doesn't? Or did I misunderstand anything?

Reply:
Yes this branch shouldn't be hit but I think we should defensively check that there is a policy when Update() is called. In the `Create()` func we handle this scenario differently and we indeed create the policy!